docs/architecture.md

```diff
@@ -0,0 +1,119 @@
+Architecture
+===
+
+Fixed-structure heterogeneous entities with no possibility of adding or removing
+components on runtime.
+
+# Entities
+Entities are structures that share a common header (preamble), as following:
+
+    Entity
+      |
+      +-- type_id: int
+      +-- flags: int
+      +-- x: int
+      +-- y: int
+      +-- userdata: ptr
+
+An example of a derived entity:
+
+    Player : Entity
+      |   ; Common fields
+      +-- type_id: int
+      +-- flags: int
+      +-- x: int
+      +-- y: int
+      +-- userdata: ptr to any
+      |   ; Player fields
+      +-- movement: MovementComponent
+      +-- health: HealthComponent
+      +-- ship: SpriteComponent
+      +-- flame: SpriteComponent
+
+
+# Components
+Components are arbitrary structures, the only thing in common they share is a
+pointer to the parent entity:
```

Comment:
Pointers to components are always stored within the entity and can also be stored in component specific lists for faster processing (eg. collision components)

Reply:
Here I mention the __pointer to entity__, sometimes in the component update function you need to refer to the entity the component belongs to, for example, for taking its position.

Reply:
yes, fully understood. My comment was more meant as extension to the current documentation. As this lists how the component stores its reference, my comment was for "where are components stored"

Reply:
You don't always need to store pointers to components in entities. Most of the time the component data is laid straight in the entity struct. If you need the reference to the component elsewhere (for example, collision detection system, that holds an array of components), you can instead have an array of pointers and register the address of the component in the entity in it.

As a general guideline, I'd suggest to always have actual component data in the struct inlined, for ease of access.

Reply:
Yes, in case the component is directly added to the entity, ofc you don't need additional pointers. Badly phrased from my side. Actually all I wanted to say was exaclty the thing you said with ", you can instead have an array of pointers and register the address of the component in the entity in it."